lens/lily/impl.go

```diff
@@ -124,3 +130,96 @@ func setupDatabase(ctx context.Context, cfg *LilyDatabaseConfig) (*storage.Datab
 }
 
 var _ LilyAPI = &LilyNodeAPI{}
+
+type HeadNotifier struct {
+	mu     sync.Mutex            // protects following fields
+	events chan *chain.HeadEvent // created lazily, closed by first cancel call
+	err    error                 // set to non-nil by the first cancel call
+}
+
+func (h *HeadNotifier) eventsCh() chan *chain.HeadEvent {
+	// caller must hold mu
+	if h.events == nil {
+		h.events = make(chan *chain.HeadEvent)
+	}
+	return h.events
+}
+
+func (h *HeadNotifier) HeadEvents() <-chan *chain.HeadEvent {
+	h.mu.Lock()
+	ev := h.eventsCh()
+	h.mu.Unlock()
+	return ev
+}
+
+func (h *HeadNotifier) Err() error {
+	h.mu.Lock()
+	err := h.err
+	h.mu.Unlock()
+	return err
+}
+
+func (h *HeadNotifier) Cancel(err error) {
+	h.mu.Lock()
+	if h.err != nil {
+		h.mu.Unlock()
+		return
+	}
+	h.err = err
+	if h.events == nil {
+		h.events = make(chan *chain.HeadEvent)
+	}
+	close(h.events)
```

Comment:
This seems odd, why make the channel here if we're just going to close it immediately?

Reply:
It's so calling cancel always results in a closed channel. A nil channel will always block in a select, but a closed one will be chosen. This means that we can have a goroutine waiting on `HeadEvents()` and as soon as the notifier is cancelled that goroutine will be notified, and this will work even if the cancel happens before the wait.